apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,283 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  History Data Adapter
+
+    The feature of History adapter is to retrieve History records from FxSync
+  Service and merge to Places DataStore which has the same functionality, and
+  the adapter is based on SyncEngine's design to handle 'update' and
+  'handleConflict' request. History Data Adapter provides Read-Only version
+  currently, and the basic ideas are included:
+  - Traverse all records from History Collection (HC).
+  - Add a new or existed place to Places DataStore (PDS).
+    - Convert the records from HC to PDS format.
+    - Merge a new record and the existed one then write to PDS.
+  - Remove an existed place when receiving a "deleted: true" record.
+**/
+
+'use strict';
+
+/* global
+  asyncStorage,
+  DataAdapters,
+  LazyLoader
+*/
+
+const HISTORY_COLLECTION_MTIME = 'collections::history::mtime';
+const SYNCTOID_PREFIX = 'SynctoId::';
+
+var HistoryHelper = (() => {
+  var placesStore;
+  function _ensureStore() {
+    if (placesStore) {
+      return Promise.resolve(placesStore);
+    }
+    return navigator.getDataStores('places').then(stores => {
+      placesStore = stores[0];
+      return Promise.resolve(placesStore);
+    });
+  }
+
+  function setSyncedCollectionMtime(mtime) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.setItem(HISTORY_COLLECTION_MTIME, mtime, resolve);
+      });
+    });
+  }
+
+  function getSyncedCollectionMtime() {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.getItem(HISTORY_COLLECTION_MTIME, resolve);
+      });
+    });
+  }
+
+  function setSynctoIdMap(synctoId, dsId) {
```

Comment:
The function name suggests this sets the entire map, not one item in the map. See also https://github.com/Kinto/kinto.js/issues/173 which means we can maybe remove this whole function in a later version.